### PR TITLE
[psc-ide] Goto definition

### PR DESCRIPTION
**Pull request by kRITZCREEK**
Works surprisingly well in my first tests. I still need to add a few integration tests, and I think the API could use some polish.

### CODE REVIEW

**Comment by kRITZCREEK** [comment]
Allright I'm happy now :) One thing that concerns me, is that I can't make the defaults assumption of `src/**/*.purs` and `bower_components...` for the sourcefiles, since we are in the compiler. This makes starting the server a bit cumbersome.

**Comment by kRITZCREEK** [comment]
Allright :) Ready for review from my side. The Windows CI is fixed and OS X are failing as always...

File: src/Language/PureScript/Ide/SourceFile.hs [hunk 1]
```diff
@@ -68,31 +66,39 @@ getImportsForFile fp = do
         unwrapImportType (P.Hiding decls)   = P.Hiding (map unwrapPositionedRef decls)
         unwrapImportType P.Implicit         = P.Implicit
 
-getDeclPosition :: P.Module -> Text -> Maybe P.SourceSpan
-getDeclPosition m ident = getFirst (foldMap (match ident) decls)
+-- | Given a surrounding Sourcespan and a Declaration from the PS AST, extracts
+-- definition sites inside that Declaration.
+extractSpans
+  :: P.SourceSpan -- ^ The surrounding span
```
Comment (paf31) [inline src/Language/PureScript/Ide/SourceFile.hs]: You might like to move the comments to the next line. The `^` is meant to point upwards to the code above, I believe.

File: src/Language/PureScript/Ide/SourceFile.hs [hunk 2]
```diff
@@ -68,31 +66,39 @@ getImportsForFile fp = do
...
+                                        -- value level one
+extractSpans ss d = case d of
+  P.PositionedDeclaration ss' _ d' ->
+    extractSpans ss' d'
+  P.ValueDeclaration i _ _ _ ->
+    [(Left (runIdentT i), ss)]
+  P.TypeSynonymDeclaration name _ _ ->
+    [(Right (runProperNameT name), ss)]
+  P.TypeClassDeclaration name _ _ members ->
+    [(Right (runProperNameT name), ss)] ++ concatMap (extractSpans' ss) members
```
Comment (paf31) [inline src/Language/PureScript/Ide/SourceFile.hs]: You can use `:` instead of `++` for some of these.

**Comment by paf31** [comment]
Looks good! So, this adds position info for type queries, but no new commands for general name queries yet, correct?

**Comment by kRITZCREEK** [comment]
Yep. I'm planning to turn the `type` command into something like ghci's `:info` to adress #1136 and #1053.

**Comment by paf31** [comment]
Could you please create an issue for this, so we have a place for comments etc.?

**Comment by paf31** [comment]
I'd like to know why Travis keeps failing on MacOS. Maybe we should just disable those builds until we can figure out what's going on.

**Comment by kRITZCREEK** [comment]
I've listed my current backlog for `psc-ide` in #2164 under *Planned Features*. I'll turn that into a checklist to show the current progress.

**Comment by kRITZCREEK** [comment]
Thanks for taking a look :) It would be nice if there was another way than me creating GIFs to showcase the new functionalities... I should either learn to hack the atom plugin so that you can try things out, or you could install a version of emacs, just to try out how the new stuff works.

I thought about making a little online playground with ace editor backed by a `psc-ide` instance to be able to play and test new features but I'll need to think about it a bit more, especially how that's supposed to work on the per-file level :)

**Comment by kRITZCREEK** [comment]
Let me know if you want me to squash.

**Comment by kRITZCREEK** [comment]
rebased and cleaned up in #2215

**Pull request closed without merging**